internal/dyngo/events.go

```diff
@@ -0,0 +1,215 @@
+package dyngo
+
+import (
+	"reflect"
+	"sync"
+	"sync/atomic"
+)
+
+type (
+	eventManager struct {
+		onStart, onData, onFinish eventRegister
+
+		disabled bool
+		mu       sync.RWMutex
+	}
+
+	eventRegister struct {
+		mu        sync.RWMutex
+		listeners eventListenerMap
+	}
+
+	// eventListenerMap is the map of event listeners. The list of listeners are indexed by the operation argument or
+	// result type the event listener expects. The map key type is a struct of the fully-qualified type name (type
+	// package path and name) to uniquely identify the argument/result type instead of using the interface type
+	// reflect.Type that can possibly lead to panics if the actual value isn't hashable as required by map keys.
+	eventListenerMap map[eventListenerMapKey][]eventListenerMapEntry
+
+	eventListenerMapKey struct {
+		pkgPath, name string
+	}
+
+	eventListenerMapEntry struct {
+		Id       eventListenerID
+		Callback interface{}
+	}
+)
+
+type (
+	EventListenerID interface {
+		unregisterFrom(*Operation)
+	}
+	OnStartEventListenerID  eventListenerID
+	OnFinishEventListenerID eventListenerID
+	OnDataEventListenerID   eventListenerID
+	eventListenerID         uint32
+)
+
+var currentEventListenerID eventListenerID
+
+func (id *eventListenerID) atomicIncrement() eventListenerID {
```

Comment:
Very good points - fixed in 335156f

Reply:
BTW, I kept the eventListenerID type definition just to be sure we don't pass accidently int values. Since the id is supposed to come out of the nextID() function, we get type-safety everywhere the ID is used.